### PR TITLE
Fix crash on unload torch cpu dll

### PR DESCRIPTION
**Pull request by DBraun**
Trying to rebase https://github.com/pytorch/pytorch/pull/61290 into latest pytorch:master

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Comment by DBraun** [comment]
Hi @ezyang, could you please help me get the Windows libtorch binaries? It looks like that kind of job wasn't triggered https://app.circleci.com/pipelines/github/pytorch/pytorch?branch=pull%2F67632

**Comment by DBraun** [comment]
Hi @jamesr66a, since you reviewed a related PR, can you take a look at this too?

**Review by jamesr66a — COMMENTED** [review]
Custom class related changes look fine to me. Adding @ilia-cher and @ZolotukhinM as reviewers for the profiler and tensorexpr changes respectively

**Comment by DBraun** [comment]
Hi @ZolotukhinM and @gdankel , would it be possible to get the libtorch binaries for this PR, the same ones that I would have otherwise gotten from pytorch.org and selecting libtorch? For me, it's more urgent to try out the binaries than for this PR to be merged. Thank you.

**Comment by ZolotukhinM** [comment]
Fuser and tensorexpr changes look good to me as well!

> would it be possible to get the libtorch binaries for this PR

You should be able to build pytorch libraries from sources: https://github.com/pytorch/pytorch#from-source

**Comment by DBraun** [comment]
Thanks for reviewing and your suggestion. I was hoping to rely on cloud infrastructure to get the binaries but I'll weigh this against compiling it locally.

**Comment by ZolotukhinM** [comment]
Ah, I see what you meant. For each PR we run lots of CI jobs and build pytorch in different configurations in the process. Maybe we can download build artifacts somehow, but I'm not sure. @malfet do you know if it's possible?

**Comment by malfet** [comment]
@DBraun binary artifacts should be available for this PR from hud.pytorch.org/pr/67632

File: torch/custom_class.h [hunk 1]
```diff
@@ -402,6 +402,23 @@ class class_ : public ::torch::detail::class_base {
     registerCustomClassMethod(std::move(method));
     return method_val;
   }
+
+  // Wrapper function to force method deregistration on shutdown.
+  static void registerCustomClassMethod(std::unique_ptr<jit::Function> method) {
```
Comment (malfet) [inline torch/custom_class.h:407]: Why couldn't we reuse RAII mechanism, already implemented in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/dispatch/RegistrationHandleRAII.h

Comment (DBraun) [reply torch/custom_class.h:407]: Just letting you know, I'm not the original author of this code, so feel free to edit as necessary.

**Comment by malfet** [comment]
Can you please add a unit-test that would crash before this change is introduced, but work fine afterwards?
Also, please note, that during dll unloading Windows runtime does not call destructors, see comment in
https://github.com/pytorch/pytorch/blob/96d116fec234e965bc735bf6b62e17942cd268d4/torch/csrc/autograd/engine.cpp#L260-L262

**Comment by DBraun** [comment]
It's pretty far out of my expertise, so I'm afraid I can't help much at this point.

**Comment by ezyang** [comment]
bumping

**Comment by DBraun** [comment]
@ezyang If you agree, please mark as not stale.

**Comment by DBraun** [comment]
Any more updates on this? Thanks and Sorry!

**Review by ezyang — APPROVED** [review]

**Comment by osalpekar** [comment]
I'm seeing a large number of test binaries crashing after this change. They seem to be primarily those using ASAN/TSAN, but also a handful that don't use such sanitizers:
```
Test appears to have passed but the binary exited with non-zero exit code -11.
This usually means something has crashed after the test was done.
Unfiltered output from the test binary:

Running main() from gmock_main.cc
Note: Google Test filter = ...
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from .../1, where TypeParam = float
[ RUN      ] .../1.NoIntersection
[       OK ] .../1.NoIntersection (0 ms)
[----------] 1 test from .../1 (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
```
Can somebody take a closer look? Meta link here: D38306879. cc @malfet @ezyang @jamesr66a

**Comment by malfet** [comment]
I propose we revert and investigate in the followup PR... Also few suggestion to followup PR:
 - Better PR description
 - Link to an issue with crash stacktrace during the unload
 - Explanation why RAII cannot/should not be used

**Comment by ezyang** [comment]
I thought the original PR's description was pretty good

**Comment by malfet** [comment]
@ezyang true, but original PR were never landed, was it?

**Comment by omid-ek** [comment]
This issue still persists in the latest release.
The crash can be reproduced with this simple snippet:
```
HMODULE h1 = LoadLibrary(L"torch_cpu.dll");
HMODULE h2 = LoadLibrary(L"torch_cuda_cu.dll");
FreeLibrary(h2);
FreeLibrary(h1);
```

**Comment by omid-ek** [comment]
update: I built pytorch by applying these changes locally and the issue was not yet resolved yet.
How can I download the check builds for this PR to test them? When I click on one of the checks all I see is "The logs for this run have expired and are no longer available."

**Pull request closed without merging**